src/Elastic.Apm/Metrics/MetricsCollector.cs

```diff
@@ -0,0 +1,124 @@
+	{
+		internal const int MaxTryWithoutSuccess = 5;
+		private readonly IApmLogger _logger;
+
+		/// <summary>
+		/// List of all providers that can provide metrics values.
+		/// Add new providers to this list in case you want the agent to collect more metrics
+		/// </summary>
+		internal readonly List<IMetricsProvider> MetricsProviders;
+		private readonly IPayloadSender _payloadSender;
+
+		private readonly Timer _timer;
+
+		public MetricsCollector(IApmLogger logger, IPayloadSender payloadSender, IConfigurationReader configurationReader)
+		{
+			_logger = logger.Scoped(nameof(MetricsCollector));
+
+			MetricsProviders = new List<IMetricsProvider>
+			{
+				new FreeAndTotalMemoryProvider(),
+				new ProcessWorkingSetAndVirtualMemoryProvider(),
+				new SystemTotalCpuProvider(_logger),
+				new ProcessTotalCpuTimeProvider()
+			};
+
+			_payloadSender = payloadSender;
+
+			logger.Debug()?.Log("starting MetricsCollector");
+
+			var interval = configurationReader.MetricsIntervalInMillisecond;
+
+			// ReSharper disable once CompareOfFloatsByEqualityOperator
+			if (interval == 0)
+			{
+				_logger.Info()?.Log("Collecting metrics is disabled - the agent won't collect metrics");
+				return;
+			}
+
+			_logger.Info()?.Log("Collecting metrics in {interval} milliseconds interval", interval);
+			_timer = new Timer(interval);
+			_timer.Elapsed += (sender, args) => { CollectAllMetrics(); };
+		}
+
+		public void StartCollecting() => _timer?.Start();
+
+		internal void CollectAllMetrics()
+		{
+			var samples = new List<Sample>();
+
+			foreach (var metricsProvider in MetricsProviders)
+			{
+				if (metricsProvider.ConsecutiveNumberOfFailedReads < MaxTryWithoutSuccess)
+				{
+					try
+					{
+						var sample = metricsProvider.GetValue();
+						if (sample != null)
+						{
+							var sampleArray = sample as Sample[] ?? sample.ToArray();
+							if (sampleArray.Any())
+								samples.AddRange(sampleArray);
+
+							metricsProvider.ConsecutiveNumberOfFailedReads = 0;
+						}
+						else
+							metricsProvider.ConsecutiveNumberOfFailedReads++;
+					}
+					catch (Exception e)
+					{
+						metricsProvider.ConsecutiveNumberOfFailedReads++;
+						_logger.Error()
+							?.LogException(e, "Failed reading {ProviderName} {NumberOfFail} times", metricsProvider.NameInLogs,
+								metricsProvider.ConsecutiveNumberOfFailedReads);
+					}
+				}
+				if (metricsProvider.ConsecutiveNumberOfFailedReads != MaxTryWithoutSuccess) continue;
+
+				_logger.Info()
+					?.Log("Failed reading {operationName} {numberOfTimes} consecutively - the agent won't try reading {operationName} anymore",
+						metricsProvider.NameInLogs, metricsProvider.ConsecutiveNumberOfFailedReads, metricsProvider.NameInLogs);
+
+				metricsProvider.ConsecutiveNumberOfFailedReads++;
+			}
+
+			var metricSet = new MetricSet(DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() * 1000, samples);
+
+			_payloadSender.QueueMetrics(metricSet);
```

Comment:
I addressed this in a commit.

Now there is a try-catch around this and in case the `PayloadSender` is already disposed then it just fails, prints a log (I'm not sure an error is good here... that's TBD, probably info is enough) and stops and disposes the timer.

I know you think this is not a good solution, nevertheless I add my opinion here: first of all there is no public feature of “stopping the agent”, once the collection started it’ll collect until the process terminates in a real environment - so this scenario is not something that happens multiple times during the life cycle of an app from a real user. On the other hand we do dispose the agent in tests, so that’s a real use case.

The solution with letting it fail is very easy to read, does not depend on any other component (you don’t have to call anything from the outside) and I believe there is not much room for bugs there - we just let it fail, handle it, and stop the timer.

So I think it solve the current challenge (which is to handle the case when the agent is disposed) well enough without high effort.

Also in case we come up with a very elegant and awesome solution, from the user’s perspective that won’t make any difference - it will be just the same when the process terminates.

I’m open for another solution, but please not spend too long time and effort on this - if there is a better solution I’m happy to take it, otherwise I’d propose to just use this approach.

Reply:
I agree - since there's no public API to stop the agent we can postpone handling the issue until we introduce such API.